File: runner.pl
#!/usr/bin/perl
use POSIX;
use Getopt::Long;
use BSD::Resource;

if(@ARGV == 0) {
  print "Usage: runner.pl <options> <command>\n";
  print "Options: \n";
  print "  --mem <bytes> - The amount of memory allowed in bytes\n";
  print "  --time <seconds> - The amount of CPU time allowed in seconds\n";
  print "  --procs <processes> - The number of processes allowed during the execution of the command\n";
  print "  --files <files> - The number of allowed open files\n";
  print "  --user <username> - The user with which to execute the command\n";
  print "  --root <path_to_root> - A directory in which to chroot the command\n";
  exit 0;
}

GetOptions("mem:i"   => \$mem,
           "time:i"  => \$time,
           "procs:i" => \$procs,
           "files:i" => \$files,
           "user:s"  => \$user,
           "root:s"  => \$root);
           
setrlimit(RLIMIT_VMEM, $mem, $mem) or die "Cannot set VMEM limit" if $mem;
setrlimit(RLIMIT_NPROC, $procs, $procs) or die "Cannot set NPROC limit" if $procs;
setrlimit(RLIMIT_OFILE, $files, $files) or die "Cannot set OFILE limit" if $files;
setrlimit(RLIMIT_CPU, $time, $time) or die "Cannot set CPU limit" if $time;

chroot($root) or die "Cannot chroot to $root" if $root;

$> = getpwnam($user) if $user;

if ($pid = fork) {
  if($time) {
    $SIG{ALRM} = sub { kill 'KILL', $pid };
    alarm 5 * $time;
  }

  waitpid($pid, 0);
  kill 'KILL', -$pid;
  
  exit($? & 127);
} else {
  setpgrp(0, $$) or die "Cannot create a process group with id $$";
  exec @ARGV;
}

File: runner.rb
#!/usr/bin/env ruby
# Runs a process with some resource limitations
# Spacial status codes:
# 127 - memory limit
# 15 - time limit
require 'getoptlong'
require 'etc'

PAGE_SIZE = 4092

def total_runtime(pid)
  # stats = File.read("/proc/#{pid}/stat").chomp.split(/\s+/)
  # stats[13].to_i + stat
  t = Process.times
  t.utime + t.stime
end

def process_state(pid)
  File.read("/proc/#{pid}/stat").chomp.split(/\s+/)[2]
end

opts = GetoptLong.new(
      [ '--help', '-h', GetoptLong::NO_ARGUMENT ],
      [ '--mem', '-m', GetoptLong::OPTIONAL_ARGUMENT ],
      [ '--time', '-t', GetoptLong::OPTIONAL_ARGUMENT ],
      [ '--procs', '-p', GetoptLong::OPTIONAL_ARGUMENT ],
      [ '--user', '-u', GetoptLong::OPTIONAL_ARGUMENT ]
    )

mem = nil
timelimit = nil
proclimit = nil
user = nil
opts.each do |opt, value|
  case opt
    when '--mem' then mem = value.to_i
    when '--time' then timelimit = value.to_i
    when '--procs' then proclimit = value.to_i
    when '--user' then user = value
  end
end

if ARGV.length < 1
  puts "No command specified to run!"
  exit 1
end

cmd = ARGV.shift

pid = fork do
  Process.setrlimit(Process::RLIMIT_CPU, timelimit, timelimit) if timelimit
  Process.setrlimit(Process::RLIMIT_NPROC, proclimit, proclimit) if proclimit
  Process.setpriority(Process::PRIO_PROCESS, 0, 20)
  
  # FIXME: the user change is not working right now
  # Process::UID.change_privilege(Etc.getpwnam(user).uid) if user
  Kernel.exec cmd
end

if !pid
  puts "Cannot start the child process! #{$!}"
  exit 1
end

loop {
  data_size = File.read("/proc/#{pid}/statm").chomp.split(/\s+/)[5]
  
  if mem and data_size.to_i * PAGE_SIZE > mem
    Process.kill "KILL", pid
    Process.waitall
    exit 127
  end
  
  if timelimit and process_state(pid) == "S"
    Process.kill "KILL", pid
    Process.waitall
    exit 9
  end
  
  status = Process.wait(pid, Process::WNOHANG)
  exit($?.exitstatus || $?.termsig || $?.stopsig) if status
  
  sleep 0.001
}

